rclrs/src/node/client.rs

```diff
@@ -0,0 +1,275 @@
+use crate::node::client::oneshot::Canceled;
+use futures::channel::oneshot;
+use std::boxed::Box;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::sync::Arc;
+
+use crate::error::{RclReturnCode, ToResult};
+use crate::MessageCow;
+use crate::Node;
+use crate::{rcl_bindings::*, RclrsError};
+
+use parking_lot::{Mutex, MutexGuard};
+use rosidl_runtime_rs::Message;
+
+// SAFETY: The functions accessing this type, including drop(), shouldn't care about the thread
+// they are running in. Therefore, this type can be safely sent to another thread.
+unsafe impl Send for rcl_client_t {}
+
+/// Internal struct used by clients.
+pub struct ClientHandle {
+    rcl_client_mtx: Mutex<rcl_client_t>,
+    rcl_node_mtx: Arc<Mutex<rcl_node_t>>,
+}
+
+impl ClientHandle {
+    pub(crate) fn lock(&self) -> MutexGuard<rcl_client_t> {
+        self.rcl_client_mtx.lock()
+    }
+}
+
+impl Drop for ClientHandle {
+    fn drop(&mut self) {
+        let handle = self.rcl_client_mtx.get_mut();
+        let rcl_node_mtx = &mut *self.rcl_node_mtx.lock();
+        // SAFETY: No preconditions for this function
+        unsafe {
+            rcl_client_fini(handle as *mut _, rcl_node_mtx as *mut _);
+        }
+    }
+}
+
+impl From<Canceled> for RclrsError {
+    fn from(_: Canceled) -> Self {
+        RclrsError::RclError {
+            code: RclReturnCode::Error,
+            msg: None,
+        }
+    }
+}
+
+/// Trait to be implemented by concrete Client structs.
+///
+/// See [`Client<T>`] for an example.
+pub trait ClientBase: Send + Sync {
+    /// Internal function to get a reference to the `rcl` handle.
+    fn handle(&self) -> &ClientHandle;
+    /// Tries to take a new response and run the callback or future with it.
+    fn execute(&self) -> Result<(), RclrsError>;
+}
+
+type RequestValue<Response> = Box<dyn FnOnce(Response) + 'static + Send>;
+
+type RequestId = i64;
+
+/// Main class responsible for sending requests to a ROS service.
+pub struct Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    pub(crate) handle: Arc<ClientHandle>,
+    requests: Mutex<HashMap<RequestId, RequestValue<T::Response>>>,
+    futures: Arc<Mutex<HashMap<RequestId, oneshot::Sender<T::Response>>>>,
+}
+
+impl<T> Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    /// Creates a new client.
+    pub fn new(node: &Node, topic: &str) -> Result<Self, RclrsError>
+    where
+        T: rosidl_runtime_rs::Service,
+    {
+        // SAFETY: Getting a zero-initialized value is always safe.
+        let mut rcl_client = unsafe { rcl_get_zero_initialized_client() };
+        let type_support = <T as rosidl_runtime_rs::Service>::get_type_support()
+            as *const rosidl_service_type_support_t;
+        let topic_c_string = CString::new(topic).map_err(|err| RclrsError::StringContainsNul {
+            err,
+            s: topic.into(),
+        })?;
+        let rcl_node = { &mut *node.rcl_node_mtx.lock() };
```

Comment:
```suggestion
        let rcl_node = &mut *node.rcl_node_mtx.lock();
```

Reply:
I can't find the thread, but this was changed based on @jhdcs 's suggestion to not lock the `Node`'s mutex for the entirety of the function and only for the handle instead. We should revisit the other handles in the codebase to make sure we're not locking too much.

Reply:
I agree with the idea, but I believe as long as you hold the mutable reference, the mutex stays locked.